code/modules/fissionreactor/datums.dm

```diff
@@ -0,0 +1,457 @@
+	else
+		totalpowertodump=fuel.wattage-fuel.absorbance
+	
+
+	if (fuel.lifetime>0)
+		fuel.life-= (speedfactor)/fuel.lifetime
+		fuel.life=max(0,fuel.life)
+	else
+		fuel.life=0
+		return
+	
+	temperature+=totalpowertodump*powerfactor/heat_capacity
+	
+
+	
+	if(temperature>=FISSIONREACTOR_MELTDOWNTEMP)
+		if(graceperiodtick)
+			meltdown()
+		graceperiodtick=TRUE
+	else
+		graceperiodtick=FALSE
+	
+	
+/datum/fission_reactor_holder/proc/coolantcycle()
+	if(coolant_ports.len==0)
+		goto skipcoolantports //avoiding div by 0. yes, this is really lazy, i know.
+	for(var/i=1, i<=coolant_ports.len,i++)
+		var/real_index= ((i+coolantport_counter)%coolant_ports.len)+1 //this way we spread out any first index prefrence.
+		var/obj/machinery/atmospherics/unary/fissionreactor_coolantport/coolant_port=coolant_ports[real_index]
+		coolant_port.transfer_reactor()
+		
+	coolantport_counter++
+	coolantport_counter=(coolantport_counter%coolant_ports.len)+1 //shift it around.	
+	
+	skipcoolantports:
+	var/chp=coolant.heat_capacity()
+	
+	//unrealistically, the heat transfer is 100% between the 2 sources.
+	//too bad.
+	var/newtemp=(chp*coolant.temperature + heat_capacity*temperature)/(chp+heat_capacity)
+	coolant.temperature=newtemp
+	temperature=newtemp
+	
+	
+	
+	
+	
+
+/datum/fission_fuel
+	var/datum/reagents/fuel= null
+	var/life=1.0 //1.0 is full life, 0 is depleted. MAKE SURE it is always 0-1 or shit WILL go wrong.
+	
+	//these are rederived when making a new one, so these can be whatever.
+	var/lifetime=0  //time in seconds that the fuel will burn for at base.
+	var/wattage=0 // heat which will be added.
+	var/absorbance=0 //subtraced from above to get total emission.
+
+/datum/fission_fuel/New()
+	fuel= new /datum/reagents //this probably isn't the best way to do things, but that's a problem for future me (someone else) to deal with.
+	//fuel.my_atom=src
+	fuel.maximum_volume=150
+
+
+	
+/datum/fission_fuel/proc/add_shit_to(var/reagent, var/amount,var/datum/reagents/holder)	//this exists because reagent code really wants an atom. but this is a datum. sux to be them. this is simpler, anyways.
+	for (var/datum/reagent/R in holder.reagent_list)
+		if (R.id == reagent)
+			R.volume += amount
+			holder.update_total()
+			return 0
+	var/datum/reagent/D = chemical_reagents_list[reagent]
+	if(D)
+		var/datum/reagent/R = new D.type()
+
+		holder.reagent_list += R
+		R.holder = holder
+		R.volume = amount
+
+		holder.update_total()
+		return 0
+	else
+		return 1
+	rederive_stats()
+	
+	
+	
+	
+/datum/fission_fuel/proc/rederive_stats() //should be called whenever you change the materials
+	if(!fuel)
+		lifetime=0
+		wattage=0
+		absorbance=0
+		return	
+	var/thislifetime=0
+	var/thiswattage=0	
+	var/thisabsorbance=0
+	
+	for(var/datum/reagent/R in fuel.reagent_list)
+		if (R.fission_time != null)
+			thislifetime+=R.fission_time* (fuel.amount_cache[R.id] + 0)/(fuel.total_volume) //fuel time is a weighted average
```

Comment:
Can fuel.total_volume ever be 0?

Reply:
only if you varedit it to be, so never in practice